File: src/Json/Decode.fs
namespace ARCtrl.Json

open Thoth.Json.Core

open ARCtrl
open Fable.Core
open System.Collections.Generic

#if FABLE_COMPILER_PYTHON
module PyTime = 

    open Fable.Core
    open Fable.Core.PyInterop

    // Currently in Fable, a created datetime object will contain a timezone. This is not allowed in python xlsx, so we need to remove it.
    // Unfortunately, the timezone object in python is read-only, so we need to create a new datetime object without timezone.
    // For this, we use the fromtimestamp method of the datetime module and convert the timestamp to a new datetime object without timezone.

    type DateTimeStatic =
        [<Emit("$0.fromtimestamp(timestamp=$1)")>]
        abstract member fromTimeStamp: timestamp:float -> System.DateTime

    [<Import("datetime", "datetime")>]
    let DateTime : DateTimeStatic = nativeOnly

    let toUniversalTimePy (dt:System.DateTime) = 
        
        dt.ToUniversalTime()?timestamp()
        |> DateTime.fromTimeStamp
#endif
    
module Helpers =

    let prependPath
        (path: string)
        (err: DecoderError<'JsonValue>)
        : DecoderError<'JsonValue>
        =
        let (oldPath, reason) = err
        (path + oldPath, reason)

    let inline prependPathToResult<'T, 'JsonValue>
        (path: string)
        (res: Result<'T, DecoderError<'JsonValue>>)
        =
        res |> Result.mapError (prependPath path)

module Decode =

    let isURI (s : string) = 
        true
        //s.StartsWith("http://") || s.StartsWith("https://")
    
    let uri : Decoder<URI> =
        { new Decoder<URI> with
             member this.Decode(s,json) = 
                    match Decode.string.Decode(s,json) with
                    | Ok s when isURI s -> Ok s
                    | Ok s -> Error (DecoderError(s,ErrorReason.FailMessage (sprintf "Expected URI, got %s" s)))
                    | Error e -> Error e
        }

    let hasUnknownFields (helpers : IDecoderHelpers<'JsonValue>) (knownFields : Set<string>) (json : 'JsonValue) = 
        helpers.getProperties json
        |> Seq.exists (fun x -> not (knownFields |> Set.contains x))

    let objectNoAdditionalProperties (allowedProperties : string seq) (builder: Decode.IGetters -> 'value) : Decoder<'value> =
        let allowedProperties = Set.ofSeq allowedProperties
        { new Decoder<'value> with
            member _.Decode(helpers, value) =
                let getters = Decode.Getters(helpers, value)
                if hasUnknownFields helpers allowedProperties value then
                    Error (DecoderError("Unknown fields in object", ErrorReason.BadPrimitive("",value)))
                else
                    let result = builder getters
                
                    match getters.Errors with
                    | [] -> Ok result
                    | fst :: _ as errors ->
                        if errors.Length > 1 then
                            ("", BadOneOf errors) |> Error
                        else
                            Error fst
        }

    let noAdditionalProperties (allowedProperties : string seq) (decoder : Decoder<'value>) : Decoder<'value> =
        let allowedProperties = Set.ofSeq allowedProperties
        { new Decoder<'value> with
            member _.Decode(helpers, value) =
                let getters = Decode.Getters(helpers, value)
                if hasUnknownFields helpers allowedProperties value then
                    Error (DecoderError("Unknown fields in object", ErrorReason.BadPrimitive("",value)))
                else
                    decoder.Decode(helpers,value)
        }

    let resizeArrayOrSingleton (decoder: Decoder<'value>) : Decoder<ResizeArray<'value>> =
        { new Decoder<ResizeArray<'value>> with
            member _.Decode(helpers, value) =
                if helpers.isArray value then
                    let mutable i = -1
                    let tokens = helpers.asArray value
                    let arr = ResizeArray()

                    (Ok arr, tokens)
                    ||> Array.fold (fun acc value ->
                        i <- i + 1

                        match acc with
                        | Error _ -> acc
                        | Ok acc ->
                            match decoder.Decode(helpers, value) with
                            | Error er ->
                                Error(
                                    er
                                    |> Helpers.prependPath (
                                        ".[" + (i.ToString()) + "]"
                                    )
                                )
                            | Ok value ->
                                acc.Add value
                                Ok acc
                    )
                else
                    decoder.Decode(helpers, value) |> Result.map (fun x -> ResizeArray[x])
        }

    let datetime: Decoder<System.DateTime> =
        #if FABLE_COMPILER_PYTHON
        Decode.datetimeLocal |> Decode.map PyTime.toUniversalTimePy
        #else
        { new Decoder<System.DateTime> with
            member _.Decode(helpers, value) =
                if helpers.isString value then
                    match System.DateTime.TryParse(helpers.asString value) with
                    | true, datetime -> datetime |> Ok
                    | _ -> ("", BadPrimitive("a datetime", value)) |> Error
                else
                    ("", BadPrimitive("a datetime", value)) |> Error
        }
        //Decode.datetimeUtc
        #endif


    let dictionary (keyDecoder : Decoder<'key>) (valueDecoder : Decoder<'value>) : Decoder<Dictionary<'key,'value>> =
        { new Decoder<Dictionary<'key,'value>> with
            member _.Decode(helpers, value) =
                if helpers.isArray value then
                    let mutable errors = []
                    let tokens = helpers.asArray value
                    let dict = Dictionary<'key,'value>()
                    let decoder = Decode.tuple2 keyDecoder valueDecoder

                    (Ok dict, tokens)
                    ||> Array.fold (fun acc value ->
                        match acc with
                        | Error _ -> acc
                        | Ok acc ->
                            match decoder.Decode(helpers, value) with
                            | Error er ->
                                Error(
                                    er                                    
                                )
                            | Ok value ->
                                acc.Add value
                                Ok acc
                    )
                    
                else
                    ("", BadPrimitive("an object", value)) |> Error
        }

    let intDictionary (valueDecoder : Decoder<'value>) : Decoder<Dictionary<int,'value>> =
        { new Decoder<Dictionary<int,'value>> with
            member _.Decode(helpers, value) =
                if helpers.isArray value then
                    let mutable errors = []
                    let tokens = helpers.asArray value
                    let dict = Dictionary<int,'value>()
                    let decoder = Decode.tuple2 Decode.int valueDecoder

                    (Ok dict, tokens)
                    ||> Array.fold (fun acc value ->
                        match acc with
                        | Error _ -> acc
                        | Ok acc ->
                            match decoder.Decode(helpers, value) with
                            | Error er ->
                                Error(
                                    er                                    
                                )
                            | Ok value ->
                                acc.Add value
                                Ok acc
                    )
                    
                else
                    ("", BadPrimitive("an object", value)) |> Error
        }

File: js/Json/Decode.js
import { Getters$2__get_Errors, Getters$2_$ctor_Z4BE6C149, string } from "../fable_modules/Thoth.Json.Core.0.4.0/Decode.fs.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.4.22.0/Result.js";
import { printf, toText } from "../fable_modules/fable-library-js.4.22.0/String.js";
import { ErrorReason$1 } from "../fable_modules/Thoth.Json.Core.0.4.0/Types.fs.js";
import { exists } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { ofSeq, contains } from "../fable_modules/fable-library-js.4.22.0/Set.js";
import { int32ToString, comparePrimitives } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { head, length, isEmpty } from "../fable_modules/fable-library-js.4.22.0/List.js";
import { fold } from "../fable_modules/fable-library-js.4.22.0/Array.js";
import { tryParse, minValue } from "../fable_modules/fable-library-js.4.22.0/Date.js";
import { FSharpRef } from "../fable_modules/fable-library-js.4.22.0/Types.js";

export function Helpers_prependPath(path, err_, err__1) {
    const err = [err_, err__1];
    return [path + err[0], err[1]];
}

export function Decode_isURI(s) {
    return true;
}

export const Decode_uri = {
    Decode(s, json) {
        const matchValue = string.Decode(s, json);
        if (matchValue.tag === 1) {
            return new FSharpResult$2(1, [matchValue.fields[0]]);
        }
        else if (Decode_isURI(matchValue.fields[0])) {
            return new FSharpResult$2(0, [matchValue.fields[0]]);
        }
        else {
            const s_3 = matchValue.fields[0];
            return new FSharpResult$2(1, [[s_3, new ErrorReason$1(6, [toText(printf("Expected URI, got %s"))(s_3)])]]);
        }
    },
};

export function Decode_hasUnknownFields(helpers, knownFields, json) {
    return exists((x) => !contains(x, knownFields), helpers.getProperties(json));
}

export function Decode_objectNoAdditionalProperties(allowedProperties, builder) {
    const allowedProperties_1 = ofSeq(allowedProperties, {
        Compare: comparePrimitives,
    });
    return {
        Decode(helpers, value) {
            const getters = Getters$2_$ctor_Z4BE6C149(helpers, value);
            if (Decode_hasUnknownFields(helpers, allowedProperties_1, value)) {
                return new FSharpResult$2(1, [["Unknown fields in object", new ErrorReason$1(0, ["", value])]]);
            }
            else {
                const result = builder(getters);
                const matchValue = Getters$2__get_Errors(getters);
                if (!isEmpty(matchValue)) {
                    const errors = matchValue;
                    return (length(errors) > 1) ? (new FSharpResult$2(1, [["", new ErrorReason$1(7, [errors])]])) : (new FSharpResult$2(1, [head(matchValue)]));
                }
                else {
                    return new FSharpResult$2(0, [result]);
                }
            }
        },
    };
}

export function Decode_noAdditionalProperties(allowedProperties, decoder) {
    const allowedProperties_1 = ofSeq(allowedProperties, {
        Compare: comparePrimitives,
    });
    return {
        Decode(helpers, value) {
            const getters = Getters$2_$ctor_Z4BE6C149(helpers, value);
            return Decode_hasUnknownFields(helpers, allowedProperties_1, value) ? (new FSharpResult$2(1, [["Unknown fields in object", new ErrorReason$1(0, ["", value])]])) : decoder.Decode(helpers, value);
        },
    };
}

export function Decode_resizeArray(decoder) {
    return {
        Decode(helpers, value) {
            if (helpers.isArray(value)) {
                let i = -1;
                return fold((acc, value_1) => {
                    let tupledArg;
                    i = ((i + 1) | 0);
                    if (acc.tag === 0) {
                        const acc_1 = acc.fields[0];
                        const matchValue = decoder.Decode(helpers, value_1);
                        if (matchValue.tag === 0) {
                            void (acc_1.push(matchValue.fields[0]));
                            return new FSharpResult$2(0, [acc_1]);
                        }
                        else {
                            return new FSharpResult$2(1, [(tupledArg = matchValue.fields[0], Helpers_prependPath((".[" + int32ToString(i)) + "]", tupledArg[0], tupledArg[1]))]);
                        }
                    }
                    else {
                        return acc;
                    }
                }, new FSharpResult$2(0, [[]]), helpers.asArray(value));
            }
            else {
                return new FSharpResult$2(1, [["", new ErrorReason$1(0, ["an array", value])]]);
            }
        },
    };
}

export const Decode_datetime = {
    Decode(helpers, value) {
        if (helpers.isString(value)) {
            let matchValue;
            let outArg = minValue();
            matchValue = [tryParse(helpers.asString(value), new FSharpRef(() => outArg, (v) => {
                outArg = v;
            })), outArg];
            return matchValue[0] ? (new FSharpResult$2(0, [matchValue[1]])) : (new FSharpResult$2(1, [["", new ErrorReason$1(0, ["a datetime", value])]]));
        }
        else {
            return new FSharpResult$2(1, [["", new ErrorReason$1(0, ["a datetime", value])]]);
        }
    },
};

